samcli/lib/utils/git_repo.py

```diff
@@ -0,0 +1,180 @@
+""" Manage Git repo """
+
+import logging
+import os
+import platform
+import shutil
+import subprocess
+from pathlib import Path
+from typing import Optional
+
+from samcli.lib.utils import osutils
+from samcli.lib.utils.osutils import rmtree_callback
+
+LOG = logging.getLogger(__name__)
+
+
+class CloneRepoException(Exception):
+    """
+    Exception class when clone repo fails.
+    """
+
+
+class CloneRepoUnstableStateException(CloneRepoException):
+    """
+    Exception class when clone repo enters an unstable state.
+    """
```

Comment:
If we want to make this `GitRepo` generic, we should provide more info about exceptions.
What does "unstable" mean here? what scenarios it cover? Does it mean "unexpected?" I think we need to add more details in this docstring.